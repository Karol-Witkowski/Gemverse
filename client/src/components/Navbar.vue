<template>
  <v-app-bar
    color="white"
    flat
    max-height="65"
  >
    <v-tabs
      app
      centered
      color="blue lighten-3"
    >
      <v-tab to="/">rooms</v-tab>
      <v-tab to="/about">about</v-tab>
    </v-tabs>
    <v-tab
      class="mt-1 mr-2"
      disabled
      style="font-size: 16px"
    >
      GEMVERSE
    </v-tab>
    <v-btn
      color="white"
      elevation="6"
      icon
      large
      outlined
      text
      to="/"
    >
      <v-img
        max-width="40"
        src="..\assets\img\logo.png"
      />
    </v-btn>
  </v-app-bar>
</template>

<script>
export default {
  name: 'Navbar',
  data() {
    return {
      drawerToggle: false,
    };
  },
};
</script>
