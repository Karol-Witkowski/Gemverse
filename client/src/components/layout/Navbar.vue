<template>
  <v-app-bar
    app
    color="white"
    flat
    max-height="65"
  >
    <v-menu
      align="center"
      justify="space-around"
    >
      <template v-slot:activator="{ on: menu, attrs }">
        <v-btn
          class="d-flex d-sm-none mx-auto"
          color="primary"
          dark
          min-width="180"
          outlined
          v-bind="attrs"
          v-on="{ ...menu }"
        >
          <v-icon>menu</v-icon>
        </v-btn>
      </template>
      <v-list>
        <v-list-item to="/">Home</v-list-item>
        <v-list-item
          to="/roomlist"
          v-if="isAuthorized"
        >
          Rooms
        </v-list-item>
        <v-list-item to="/about">About</v-list-item>
        <v-list-item v-if="isAuthorized">
          <v-btn
            class="mx-auto"
            @click.prevent="logout"
            color="primary"
            outlined
            small
            width="250"
          >
            Logout
          </v-btn>
        </v-list-item>
      </v-list>
    </v-menu>
    <v-tabs class="hidden-xs-only" color="primary">
      <v-tab to="/">Home</v-tab>
      <v-tab
        to="/roomlist"
        v-if="isAuthorized"
      >
        Rooms
      </v-tab>
      <v-tab to="/about">About</v-tab>
    </v-tabs>
    <v-tab to="/">
      <span class="appName grey--text text--darken-2">Gemverse</span>
    </v-tab>
    <v-tab
      class="hidden-xs-only mx-2"
      @click.prevent="logout"
      v-if="isAuthorized"
    >
      Logout
    </v-tab>
  </v-app-bar>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'Navbar',

  created() {
    this.dispatchToken();
  },

  computed: {
    ...mapActions(['remitAuthState']),
    ...mapGetters(['isAuthorized']),
  },

  methods: {
    dispatchToken() {
      if (localStorage.getItem('authenticationToken')) {
        this.$store.dispatch('remitAuthState', true);
      } else {
        localStorage.clear();
        this.$store.dispatch('remitAuthState', false);
      }
    },

    logout() {
      if (localStorage.getItem('authenticationToken')) {
        localStorage.clear();
        this.$store.dispatch('remitAuthState', false);
        this.$store.dispatch('saveUser', '');
        this.$router.push({ name: 'Login' });
      }
    },
  },
};
</script>
<style lang="scss">
.appNavigation {
  font-size: 1px!important;
  font-weight: 600;
}

.appName {
  letter-spacing: 4px;
}
</style>
