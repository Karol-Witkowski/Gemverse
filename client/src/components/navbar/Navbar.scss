nav {
  padding: 30px;

  a {
    font-weight: bold;
    color: #2d7c58;

    &.router-link-exact-active {
      color: #42b983;
    }
  }
}