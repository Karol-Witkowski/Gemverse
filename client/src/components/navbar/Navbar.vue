<template>
  <nav>
    <router-link to="/">Main</router-link> |
    <router-link to="/about">About</router-link>
  </nav>
</template>

<script>
export default {
  name: 'Navbar',
};
</script>

<style src="./Navbar.scss" scoped lang="scss" />
