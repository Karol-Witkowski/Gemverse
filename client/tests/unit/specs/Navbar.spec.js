import { createLocalVue, mount } from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Navbar from '@/components/layout/Navbar.vue';

const localClear = jest.spyOn(window.localStorage.__proto__,
  'clear',
);
const localVue = createLocalVue();
let vuetify;
let wrapper;

vuetify = new Vuetify();

describe('Implementation test for Navbar.vue - unauthorized user', () => {
  const mockStoreUnauth = {
    dispatch: jest.fn(),
    getters: {
      isAuthorized: false,
    },
  };

  beforeEach(() => {
    wrapper = mount(Navbar, {
      localVue,
      mocks: {
        $store: mockStoreUnauth,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    expect(wrapper.findAll('.v-tab').length).toEqual(3);
    expect(wrapper.findAll('.v-tab').at(0).text()).toMatch('home');
    expect(wrapper.findAll('.v-tab').at(1).text()).toMatch('about');
    expect(wrapper.findAll('.v-tab').at(2).text()).toMatch('gemverse');

    // Check that mobile menu is rendered
    expect(wrapper.findAll('button').length).toEqual(1);
    expect(wrapper.findAll('button').at(0).text()).toMatch('menu');
  });

  it('Should call dispatch method when created', () => {
    // Check if any action were dispatched
    expect(mockStoreUnauth.dispatch).toHaveBeenCalled();

    // Check if one action was dispatched
    expect(mockStoreUnauth.dispatch).toHaveReturnedTimes(1);
  });

  it('Dispatch correct auth state when created', () => {
    expect(mockStoreUnauth.dispatch).toHaveBeenCalledWith(
      'remitAuthState',
      false,
    );
  });

  it('Should call store clear method when created', () => {
    // Check if local storage clear method was called
    expect(localStorage.clear).toHaveBeenCalled();

    // Check if local storage clear method was called once
    expect(localStorage.clear).toHaveReturnedTimes(1);
  });
});

describe('Implementation test for Navbar.vue - authorized user', () => {
  const localGetItem = jest.spyOn(window.localStorage.__proto__,
    'setItem',
  );
  const mockRouter = {
    push: jest.fn()
  };
  const mockStoreAuth = {
    dispatch: jest.fn(),
    getters: {
      isAuthorized: true,
    },
  };

  beforeEach(() => {
    localStorage.setItem('authenticationToken', 'testToken');

    wrapper = mount(Navbar, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStoreAuth,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    localGetItem.mockClear();
    wrapper.destroy();
  });

  it('Initializes with correct elements', () => {
    // Check tabs initial state
    expect(wrapper.findAll('.v-tab').length).toEqual(6);
    expect(wrapper.findAll('.v-tab').at(0).text()).toMatch('home');
    expect(wrapper.findAll('.v-tab').at(1).text()).toMatch('rooms');
    expect(wrapper.findAll('.v-tab').at(2).text()).toMatch('profile');
    expect(wrapper.findAll('.v-tab').at(3).text()).toMatch('about');
    expect(wrapper.findAll('.v-tab').at(4).text()).toMatch('gemverse');
    expect(wrapper.findAll('.v-tab').at(5).text()).toMatch('logout');

    // Check that mobile menu is rendered
    expect(wrapper.findAll('button').length).toEqual(1);
    expect(wrapper.findAll('button').at(0).text()).toMatch('menu');
  });

  it('Should call dispatch method when created', () => {
    // Check if any action were dispatched
    expect(mockStoreAuth.dispatch).toHaveBeenCalled();

    // Check if one action was dispatched
    expect(mockStoreAuth.dispatch).toHaveReturnedTimes(1);
  });

  it('Dispatch correct auth state when created', () => {
    expect(mockStoreAuth.dispatch).toHaveBeenCalledWith(
      'remitAuthState',
      true,
    );
  });

  it('Dispatch correct data on logout', () => {
    wrapper.vm.logout();

    // Check if any action were dispatched
    expect(mockStoreAuth.dispatch).toHaveBeenCalled();

    // Check if two actions were dispatched
    expect(mockStoreAuth.dispatch).toHaveReturnedTimes(3);

    // Check auth state was dispatched with correct data
    expect(mockStoreAuth.dispatch).toHaveBeenNthCalledWith(
      1,
      'remitAuthState',
      true,
    );

    // Check user data was dispatched with correct data
    expect(mockStoreAuth.dispatch).toHaveBeenNthCalledWith(
      3,
      'saveUser',
      '',
    );
  });

  it('Add auth state to storage object', () => {
    wrapper.vm.logout();

    // Check if local storage setter was called
    expect(localStorage.setItem).toHaveBeenCalled();

    // Check if local storage setter was called once
    expect(localStorage.setItem).toHaveReturnedTimes(1);

    // Check if local storage setter was called with correct data
    expect(localStorage.setItem).toHaveBeenCalledWith(
      'authenticationToken',
      'testToken',
    );
  });

  it('Clear local storage on logout', () => {
    wrapper.vm.logout();

    // Check if local storage clear method was called
    expect(localStorage.clear).toHaveBeenCalled();

    // Check if local storage clear method was called once
    expect(localStorage.clear).toHaveReturnedTimes(1);
  });

  it('Redirect user to login page on logout', () => {
    wrapper.vm.logout();

    // Check if router push was called
    expect(mockRouter.push).toHaveBeenCalled();

    // Check if router push was called one time
    expect(mockRouter.push).toHaveBeenCalledTimes(1)

    // Check if router push was called with correct object
    expect(mockRouter.push).toHaveBeenCalledWith({
      "name": "Login"
    });
  });
});
