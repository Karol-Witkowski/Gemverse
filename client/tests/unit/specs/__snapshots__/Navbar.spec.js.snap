// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for Navbar.vue - unauthorized user Render correctly 1`] = `
<header class="v-sheet theme--light v-toolbar v-toolbar--flat v-app-bar v-app-bar--fixed white" style="max-height: 65px; height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;">
  <div class="v-toolbar__content" style="height: 64px;"><button type="button" class="d-flex d-md-none mx-auto v-btn v-btn--outlined theme--dark v-size--small primary--text" style="width: 200px;" name="menu" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content"><i aria-hidden="true" alt="hamburger menu" class="v-icon notranslate material-icons theme--dark">menu</i></span></button>
    <div class="v-menu" align="center" justify="space-around">
      <!---->
    </div>
    <div class="spacer"></div>
    <div class="v-tabs d-none d-md-block theme--light">
      <div role="tablist" class="v-item-group theme--light v-slide-group v-tabs-bar v-tabs-bar--is-mobile primary--text">
        <div class="v-slide-group__prev v-slide-group__prev--disabled">
          <transition-stub name="fade-transition"></transition-stub>
        </div>
        <div class="v-slide-group__wrapper">
          <div class="v-slide-group__content v-tabs-bar__content">
            <div class="v-tabs-slider-wrapper" style="transition: none;">
              <div class="v-tabs-slider"></div>
            </div>
            <router-link-stub tabindex="0" aria-selected="true" role="tab" class="v-tab">home</router-link-stub>
            <!---->
            <!---->
            <router-link-stub tabindex="0" aria-selected="false" role="tab" class="v-tab">about</router-link-stub>
          </div>
        </div>
        <div class="v-slide-group__next v-slide-group__next--disabled">
          <transition-stub name="fade-transition"></transition-stub>
        </div>
      </div>
    </div>
    <router-link-stub tabindex="0" aria-selected="false" role="tab" class="appName grey--text menu text--darken-2 v-tab">
      gemverse
    </router-link-stub>
    <!---->
  </div>
</header>
`;
